sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala

```diff
@@ -609,20 +544,62 @@ private[sql] object DataSourceV2Strategy {
   }
 
   protected[sql] def rebuildExpressionFromFilter(
-      filter: V2Filter,
-      translatedFilterToExpr: mutable.HashMap[V2Filter, Expression]): Expression = {
-    filter match {
+      predicate: Predicate,
+      translatedFilterToExpr: mutable.HashMap[Predicate, Expression]): Expression = {
+    predicate match {
       case and: V2And =>
-        expressions.And(rebuildExpressionFromFilter(and.left, translatedFilterToExpr),
-          rebuildExpressionFromFilter(and.right, translatedFilterToExpr))
+        expressions.And(
+          rebuildExpressionFromFilter(and.left(), translatedFilterToExpr),
+          rebuildExpressionFromFilter(and.right(), translatedFilterToExpr))
       case or: V2Or =>
-        expressions.Or(rebuildExpressionFromFilter(or.left, translatedFilterToExpr),
-          rebuildExpressionFromFilter(or.right, translatedFilterToExpr))
+        expressions.Or(
+          rebuildExpressionFromFilter(or.left(), translatedFilterToExpr),
+          rebuildExpressionFromFilter(or.right(), translatedFilterToExpr))
       case not: V2Not =>
-        expressions.Not(rebuildExpressionFromFilter(not.child, translatedFilterToExpr))
-      case other =>
-        translatedFilterToExpr.getOrElse(other,
-          throw new IllegalStateException("Failed to rebuild Expression for filter: " + filter))
+        expressions.Not(rebuildExpressionFromFilter(not.child(), translatedFilterToExpr))
+      case _ =>
+        translatedFilterToExpr.getOrElse(predicate,
+          throw new IllegalStateException("Failed to rebuild Expression for filter: " + predicate))
+    }
+  }
+}
+
+abstract class PushablePredicateBase {
+  val nestedPredicatePushdownEnabled: Boolean
+  val pushableColumn: PushableColumnBase = PushableColumn(nestedPredicatePushdownEnabled)
+
+  def unapply(e: Expression): Option[Predicate] = e match {
+    case expressions.Literal(true, BooleanType) =>
+      Some(new Predicate("TRUE", Array.empty))
+    case expressions.Literal(false, BooleanType) =>
+      Some(new Predicate("FALSE", Array.empty))
+    case col @ pushableColumn(name) if col.dataType.isInstanceOf[BooleanType] =>
+      Some(new Predicate(name, Array.empty))
+    case _ =>
```

Comment:
It doesn't seem necessary to have the above 3 cases. Even if we guarantee boolean type here, we may still hit `Add(a, b)` where `a` and `b` are not boolean type columns. Let's just call `V2ExpressionBuilder`

Reply:
This is the outermost filter

Reply:
If we address https://github.com/apache/spark/pull/35768/files#r825606670 , we don't need code to handle out-most filter.